src/segment_evaluation.rs

```diff
@@ -19,125 +19,156 @@ use crate::{
     entity::{AttrValue, Entity},
     models::TargetingRule,
 };
+use crate::errors::{Result, Error};
+
+// For chaining errors creating useful error messages:
+use anyhow::{Context, Result as AnyhowResult};
+use anyhow::anyhow;
 
 pub(crate) fn find_applicable_segment_rule_for_entity(
     segments: &HashMap<String, Segment>,
     segment_rules: impl Iterator<Item = TargetingRule>,
     entity: &impl Entity,
-) -> Option<TargetingRule> {
+) -> Result<Option<TargetingRule>> {
     let mut targeting_rules = segment_rules.collect::<Vec<_>>();
     targeting_rules.sort_by(|a, b| a.order.cmp(&b.order));
-    targeting_rules
-        .into_iter()
-        .find(|targeting_rule| targeting_rule_applies_to_entity(segments, targeting_rule, entity))
+
+    for targeting_rule in targeting_rules.into_iter(){
```

Comment:
I failed to find a nice way of propagating errors out of iterator chains. So in many cases like this I needed to fall-back to explicit loops. Any ideas?

Reply:
https://github.com/rainerschoe/appconfiguration-rust-sdk/pull/1